### ISSUE TITLE
Poorly optimized for processing many images

### ISSUE THREAD

**Issue by oxysoft** [chaiNNer-org/chaiNNer#2279]
**Motivation**
Hi, currently it will take **2 days** to finish rendering, and I have 20 GB of VRAM sitting unused. Why isn't ChaiNNer using this to process more images? This looks like a mature project so it's really disappointing to see that I have to revert back to writing code and doing things myself.

**Description**
Must implement batch processing in iterators.

**Alternatives**

**Comment by RunDevelopment** [comment]
Please provide more information. Right now, I have no idea what it is that you are doing. I only know 1) that you are using some iterator (but not which one), 2) that you are doing some image processing, and 3) that it's slow. 

Please share a screenshot of your chain, so we know *what* is slow and can figure out *why* it is slow.

**Comment by oxysoft** [comment]
Well I dragged a video file with 30,000 frames and used one of the recent SRFormer model, and it processes about 1 frame per 5 second. Not much else to say, looked something like this

![image](https://github.com/chaiNNer-org/chaiNNer/assets/3945277/9c74fd54-ae92-4aa7-9e9d-b3ae71480961)

**Comment by joeyballentine** [comment]
It doesn't instantiate the model 30,000 times. It caches those nodes. If that's not the behavior you're seeing (and you can prove that it isn't caching) please file a separate bug report for that. 

As for multithreading, that's pretty difficult given how chaiNNer works. We might be able to accomplish that with a threads parameter and the new iterator system you're working on though. Generally speaking though we can't just automatically do multithreading for chains as things need to run in order, and for iteration most people don't have a xx90 GPU and would easily go out of vram with it, so it hasn't been a priority to implement.

**Comment by RunDevelopment** [comment]
Ah, video frame iterator. Yeah, that iterator is *really* slow right now. Basically, we currently 1) load one frame, 2) process it (upscaling in your case) and 3) write the processed frame to disk, all in sequence. This is obviously inefficient, because we aren't even loading the next frame while writing the current processed one.

Well, the new iterator system should make it a lot easier to optimize this (and other) iterators, so look forward to some performance improvements in the next release.

**Comment by oxysoft** [comment]
Even on my previous card with 6GB, with some models I could run 4-5 of them in parallel for basically a 500% performance boost, so it's definitely worth it. I wasn't saying ChaiNNer is loading 30,000 of them, I just thought if you put the Load Model node inside the iterator then it would load for each iteration. I'm really looking forward to the iterator upgrades, ChaiNNer is A+ in every other way!

**Comment by joeyballentine** [comment]
> I was just under the assumption that if you put the Load Model node inside the iterator then it would load for each iteration

Nope. We optimize those so that they run as if they were outside the iterator.

> Even on my previous card with 6GB, with some models I could run 4-5 of them in parallel for basically a 500% performance boost, so it's definitely worth it.

That's still highly dependent on the model and the resolution of the content you're upscaling. A regular ESRGAN model at around 720p basically maxes out my vram.

My philosophy around the upscaling in chaiNNer is that users shouldn't need to worry about their vram usage, which is why we have automatic tiling to avoid OOMs. I'm not a huge fan of adding a feature that a new user could get confused by and footgun themselves into always running out of memory by using too many parallel threads. I'd much rather have some kind of automatic threading with a GPU vram-lock queue thing where it tries to automatically parallelize as much as it can automatically without causing issues. That's just a very difficult problem to solve, and I am by no means an expert in solving those kinds of problems.

**Comment by Milor123** [comment]
> > Estaba asumiendo que si colocas el nodo Cargar modelo dentro del iterador, se cargaría para cada iteración.
> 
> No. Los optimizamos para que se ejecuten como si estuvieran fuera del iterador.
> 
> > Incluso en mi tarjeta anterior con 6GB, con algunos modelos podía ejecutar 4-5 de ellas en paralelo para obtener básicamente un aumento de rendimiento del 500%, por lo que definitivamente vale la pena.
> 
> Eso todavía depende en gran medida del modelo y la resolución del contenido que estás ampliando. Un modelo ESRGAN normal a alrededor de 720p básicamente maximiza mi vram.
> 
> Mi filosofía sobre la mejora en chaiNNer es que los usuarios no deberían tener que preocuparse por el uso de su vram, razón por la cual tenemos mosaicos automáticos para evitar OOM. No soy un gran admirador de agregar una característica que pueda confundir a un nuevo usuario y quedarse siempre sin memoria al usar demasiados subprocesos paralelos. Preferiría tener algún tipo de subprocesamiento automático con una cola de bloqueo de vram de GPU donde intente paralelizar automáticamente tanto como pueda sin causar problemas. Ése es simplemente un problema muy difícil de resolver y de ninguna manera soy un experto en resolver ese tipo de problemas.

Maybe could try do in a experimental feature, in the which the user can select aprox the VRAM and brand (we know that AMD is a shit for the VRAM management) and according to this try allow some parellel threads or something similar. For images upscaly its could be very good, i feel the process very very slow with the "image per image mode"

**Comment by markcagatandavis** [comment]
Just adding to this, if you want me to create a new issue I will. I have been doing testing and noticed that the times are really slow... For my testing (not this happens with other models) I used REAL-ESRGAN and the model 'realesr-animevideov3' across 34,799 frames animation.

REAL-ESRGAN takes 2 hours to upscaled all frames: 
```
$RealsrganExePath = "P:\realesrgan-ncnn-vulkan-20220424-windows\realesrgan-ncnn-vulkan.exe"
$SrcFrames = "P:\PROCESSING\src_frames"
$UpscaledFrames = "P:\PROCESSING\upscaled_frames"
$Module = "P:\realesrgan-ncnn-vulkan-20220424-windows\module\realesr-animevideov3-x4"

"& '$RealsrganExePath' -i '$SrcFrames' -o '$UpscaledFrames' -n '$Module'"
```

chaiNNer it takes 11 hours for the same thing:
`Load Images (recursive) -> Upscale (default x4) -> Save Image (same name + sub directory). `
Upscale takes 0.23seconds per-frame

I am not sure of the differences between the REAL-ESRGAN.exe and chaiNNer when it comes to processing... But I did test this with Upscayl, stable diffusion, and some custom scripts and they all do them in significantly less time. 

I'm happy to provide additional testing, logs, results if that will help :) 
I love everything about this project! Such an amazing job, it inspired me to learn python and as a fun project I am learning to make my own UI and things for FFMPEG and upscaling. (still many years to go, but thank you)
